### ISSUE TITLE
Initial startup error - Error executing command: java.io.IOException: Stream Closed

### ISSUE THREAD

**Issue by martinvw** [openhab/openhab-distro#573]
When starting a clean install the first time it starts the karaf console locks down and reports an error as soon as you press enter:

```
Error executing command: java.io.IOException: Stream Closed
```

## Initial investigation

After pressing enter I see some NPE's in [karaf's SecuredSessionFactoryImpl](https://github.com/apache/karaf/blob/karaf-4.1.x/shell/core/src/main/java/org/apache/karaf/shell/impl/console/osgi/secured/SecuredSessionFactoryImpl.java#L217) but looking at the code they should be ignored `// Ignore, should never happen`

The strange thing is that while the console prints 'java.io.IOException' I see no such exception when linking a debugger to the process.

Another exception which I see very often is that it tries to check whether the README in the addons folder is a zip file, but that one also seems to be properly handled.

## Reproducing

By just removing all folders from `userdata` except for `tmp` and `etc` this can be reliably reproduced.

## Additional information

When the system is in that state it also does not claim the usual ports:

```
laptop-106359:openhab mvanwinger$ sudo lsof -nPi | grep ` ps aux | grep openhab | grep java | awk '{ print $2 }'`
java      14728     mvanwinger    5u  IPv4 0x8bfcd3076c95ccc9      0t0    TCP *:5005 (LISTEN)
java      14728     mvanwinger   37u  IPv6 0x8bfcd30734e3f369      0t0    TCP 127.0.0.1:53439 (LISTEN)
```

But in the normal state (second boot):

```
laptop-106359:openhab mvanwinger$ sudo lsof -nPi | grep ` ps aux | grep openhab | grep java | awk '{ print $2 }'`
java      14317     mvanwinger    5u  IPv4 0x8bfcd3076c95d5c1      0t0    TCP *:5005 (LISTEN)
java      14317     mvanwinger   30u  IPv6 0x8bfcd3072f7c48a9      0t0    TCP 127.0.0.1:53033 (LISTEN)
java      14317     mvanwinger   53u  IPv6 0x8bfcd3073179f369      0t0    TCP *:8080 (LISTEN)
java      14317     mvanwinger   88u  IPv6 0x8bfcd307377dc7c9      0t0    UDP *:1900
java      14317     mvanwinger   89u  IPv6 0x8bfcd307377de1e9      0t0    UDP *:1900
java      14317     mvanwinger   90u  IPv6 0x8bfcd307377dc309      0t0    UDP 193.0.22.136:62548
java      14317     mvanwinger   91u  IPv6 0x8bfcd307377ddd29      0t0    UDP 192.168.1.208:62656
java      14317     mvanwinger  204u  IPv6 0x8bfcd3072f7c4369      0t0    TCP 127.0.0.1:8101 (LISTEN)
java      14317     mvanwinger  235u  IPv6 0x8bfcd30734352369      0t0    TCP *:8443 (LISTEN)
java      14317     mvanwinger  300u  IPv6 0x8bfcd307377dd609      0t0    UDP *:5353
java      14317     mvanwinger  301u  IPv6 0x8bfcd307377d9849      0t0    UDP *:5353
java      14317     mvanwinger  302u  IPv6 0x8bfcd307377d8549      0t0    UDP *:5353
```
 
PS: the port 5005 is the debugger port

**Comment by martinvw** [comment]
@kaikreuzer see above, I still fear that this might the reason that docker(-compose) support seems broken.

**Comment by kaikreuzer** [comment]
Thanks for debugging this - I would assume that the root cause is some where in Karaf, unfortunately, it does not happen on a "plain" Karaf installation, so we will have to figure out, which part of openHAB provokes it.

**Comment by martinvw** [comment]
Does the high numbered port on which it is listening ring any bell on your side?

**Comment by kaikreuzer** [comment]
Not at all.
Note that Karaf is using JLine for Console I/O, which has been [updated to 3.5.0](https://issues.apache.org/jira/browse/KARAF-5359) in Karaf 4.1.3.
I would very much assume the exceptions coming from that library.

**Comment by martinvw** [comment]
That port is also stored in '/userda/tmp/port', so that is a hint

**Comment by kaikreuzer** [comment]
Interesting indeed!

**Comment by martinvw** [comment]
Or not that interesting the shutdown thread is also there when it is properly started (port 53033 in the output above). So the only thing we can learn is that karaf does not start but only the shutdown thread is properly started. ...

**Comment by splatch** [comment]
I didn't run into yet as my setup is still on OH21, but based on your inputs it clearly looks like shell issue. To confirm that console stuff is guilty one you can try to set up start local shell via adding -Dkaraf.startLocalConsole=false -Dkaraf.startRemoteShell=false to Karaf args.
Jline and console stuff is also quite sensitive on TERM settings, you might try to set this environment variable to most basic one.

I would rather bet on some shell stuff as it says "error executing command", however this is most likely some poor log or system out/err call which logs only exception message. You can also try with shell aliases and removing anything which is specific to openhab. If plain karaf 4.1.3 works in your environment than we just need to find configuration which is handling got broken.

**Comment by martinvw** [comment]
Fine tuning the logging configuration gave us no surprise ;-)

The real exception we are facing is the following:

```
openhab> 07:40:00.581 [ERROR] [apache.karaf.shell.support.ShellUtil] - Exception caught while executing command
java.io.IOError: java.io.IOException: Stream Closed
	at org.jline.keymap.BindingReader.readCharacter(BindingReader.java:142) ~[14:org.jline:3.5.0]
	at org.jline.keymap.BindingReader.readBinding(BindingReader.java:109) ~[14:org.jline:3.5.0]
	at org.jline.keymap.BindingReader.readBinding(BindingReader.java:60) ~[14:org.jline:3.5.0]
	at org.jline.reader.impl.LineReaderImpl.readBinding(LineReaderImpl.java:724) ~[14:org.jline:3.5.0]
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:526) ~[14:org.jline:3.5.0]
	at org.apache.karaf.shell.impl.console.ConsoleSessionImpl.run(ConsoleSessionImpl.java:349) [11:org.apache.karaf.shell.core:4.1.3]
	at java.lang.Thread.run(Thread.java:745) [?:?]
Caused by: java.io.IOException: Stream Closed
	at java.io.FileInputStream.read0(Native Method) ~[?:?]
	at java.io.FileInputStream.read(FileInputStream.java:207) ~[?:?]
	at org.jline.terminal.impl.DumbTerminal$1.read(DumbTerminal.java:48) ~[?:?]
	at org.jline.terminal.impl.DumbTerminal$1.read(DumbTerminal.java:85) ~[?:?]
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284) ~[?:?]
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326) ~[?:?]
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178) ~[?:?]
	at sun.nio.cs.StreamDecoder.read0(StreamDecoder.java:127) ~[?:?]
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:112) ~[?:?]
	at java.io.InputStreamReader.read(InputStreamReader.java:168) ~[?:?]
	at org.jline.utils.NonBlockingReader.run(NonBlockingReader.java:276) ~[?:?]
	... 1 more
Error executing command: java.io.IOException: Stream Closed
```

**Comment by martinvw** [comment]
@splatch disabling the the local and remote shell in the `karaf` script it indeed restores functioning without a restart:

https://github.com/openhab/openhab-distro/blob/master/distributions/openhab/src/main/resources/bin/karaf#L149

```
OPTS="-Dkaraf.startLocalConsole=false -Dkaraf.startRemoteShell=false"
```

And while the first time it starts up a dump terminal the second time it succeeds:

```
2017-11-21 08:17:27.596 [DEBUG] [org.jline                           ] - Registering shutdown-hook: Thread[JLine Shutdown Hook,5,main]
2017-11-21 08:17:27.597 [DEBUG] [org.jline                           ] - Adding shutdown-hook task: org.jline.terminal.impl.PosixSysTerminal$$Lambda$31/742664625@7fc4c418
2017-11-21 08:17:27.597 [DEBUG] [org.jline                           ] - Using terminal PosixSysTerminal
2017-11-21 08:17:27.598 [DEBUG] [org.jline                           ] - Using pty OsXNativePty
```

**Comment by splatch** [comment]
@martinvw I can reproduce issue with OH but not Karaf 4.1.3 alone. From error it seems there is a some issue with JNA.
```Caused by: java.lang.ClassNotFoundException: com.sun.jna.Platform cannot be found by org.jline_3.5.0```

Will take a look on it.

**Comment by martinvw** [comment]
On the second boot it seems available could it be caused by the startup error?

**Comment by splatch** [comment]
I did quick run and I think JNA is just some kind of side effect. It is not imported via system classpath and it is optional. What I have found is that jline tries to execute `tty` command. Command is found but reports "not a tty" at first run. Same thing happens for me on linux.

**Comment by wborn** [comment]
The issue seems to be fixed for me when I use [jline-3.5.1.jar](http://search.maven.org/remotecontent?filepath=org/jline/jline/3.5.1/jline-3.5.1.jar) instead of jline-3.5.0.jar.

E.g. test it by mounting the JAR into a container like:
```
-v "./jline-3.5.1.jar:/openhab/runtime/system/org/jline/jline/3.5.0/jline-3.5.0.jar"
```

**Comment by splatch** [comment]
Yup, I can confirm it solves issue on OSX too.

**Comment by martinvw** [comment]
Did one of you see a specific commit in jline which could solve/explain this?

**Comment by wborn** [comment]
@martinvw I can't pin point it to a single commit. 

Karaf 4.2.0 will also include the version 3.5.1 of jline, see  [pom.xml (master)]( https://github.com/apache/karaf/blob/75f345cbf990134f9dad551591380e76d887bb36/pom.xml#L260).

**Comment by kaikreuzer** [comment]
Probably rather too late for us as I would like to do the openHAB 2.2 release still this year.
So I could try to force 3.5.1 into our distro nonetheless - as I just did something similar for Jetty, I hope I can manage to do so...

**Comment by splatch** [comment]
We could stay on current version of Karaf and just use overrides mechanism to force patched version of jline. This would be work around which allows to proceed without additional release from Karaf side.

**Comment by maggu2810** [comment]
@splatch You refer to overrides.properties. Correct?

**Comment by kaikreuzer** [comment]
There's always still something to learn. I would not touch the Jetty stuff now anymore, but would one of you that is familiar with this mechanism come up with a PR to include the patched JLine lib?

**Comment by maggu2810** [comment]
For Jetty it would be for example `overrides.properties`:

```text
mvn:org.eclipse.jetty.http2/http2-common/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.http2/http2-hpack/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.http2/http2-server/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-alpn-server/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-client/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-continuation/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-deploy/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-http/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-io/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-jaas/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-jaspi/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-jmx/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-jndi/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-plus/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-rewrite/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-security/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-server/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-servlet/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-servlets/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-util/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-util-ajax/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-webapp/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty/jetty-xml/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.osgi/jetty-osgi-alpn/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/javax-websocket-client-impl/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/javax-websocket-server-impl/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/websocket-api/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/websocket-client/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/websocket-common/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/websocket-server/9.3.22.v20171030;range=[9.3.0,9.4)
mvn:org.eclipse.jetty.websocket/websocket-servlet/9.3.22.v20171030;range=[9.3.0,9.4)
```

The range could perhaps be dropped, but this is a little bit more explicit.

**Comment by splatch** [comment]
> @splatch You refer to overrides.properties. Correct?
maggu2810 Yeah, that's thing which will allow to just bump jline alone. Overrides forces karaf features to use higher version than specified (and this is important - it needs to be higher in terms of OSGi versioning, so 1.1.0 can be replaced even by 1.1.0-a). Library itself still needs to be present in system directory or any online repository configured in pax url mvn, meaning there is a need to maintain separate "feature" file which embeds patched JARs with distro.

**Comment by kaikreuzer** [comment]
Anyone willing to create a PR?

**Comment by maggu2810** [comment]
Let's give it a try:
https://github.com/openhab/openhab-distro/compare/master...maggu2810:overrides?expand=1

But building and testing takes some time...

**Comment by maggu2810** [comment]
JLine 3.5.1 will be used.

```text
openhab> bundle:list -t 0 -l | grep jline
 14 │ Active   │  30 │ 3.5.1                  │ mvn:org.jline/jline/3.5.1
```

Now 3.5.0 and 3.5.1 are part of the system directory. Because 3.5.0 is bundled because of the standard feature and 3.5.1 because of the overrides.
If you don't want to include both because you want to drop some bytes, you need to add an exclude to your assembly XML instruction file, too.

**Comment by kaikreuzer** [comment]
Fixed by https://github.com/openhab/openhab-distro/pull/577.

Thanks @maggu2810!

**Comment by kaikreuzer** [comment]
> you need to add an exclude to your archive XML instruction file

Done: https://github.com/openhab/openhab-distro/pull/578

**Comment by kaikreuzer** [comment]
#577 has been reverted, so I am afraid we have to reopen this issue.

**Comment by kaikreuzer** [comment]
So what are our options now? Short summary of where we stand:
- Karaf 4.1.3 is not compatible with JLine 3.5.1
- Karaf 4.2.0.M1 is compatible with JLine 3.5.1, but has a [severe bug](https://github.com/openhab/openhab-distro/commit/6c1614fddf22b66d0674b02969405dff3189c182#commitcomment-25800760), which is only [fixed on master](https://github.com/openhab/openhab-distro/commit/6c1614fddf22b66d0674b02969405dff3189c182#commitcomment-25800795).
- Karaf 4.1.4 is scheduled for mid December, 4.2.0.M2 for end December only.

All not very satisfying :-(
Might it be worth a try to downgrade JLine to 3.4.0 and see if that works?

**Comment by kaikreuzer** [comment]
Argh, JLine 3.4.0 isn't a good idea either:
```
Unable to resolve org.apache.karaf.shell.core/4.1.3: missing requirement [org.apache.karaf.shell.core/4.1.3] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.jline.builtins)(version>=3.5.0)(!(version>=4.0.0)))"]]
```

**Comment by martinvw** [comment]
I don't know the release cycle of JLine, but maybe we can do a pull request to include a deprecated constructor to be backwards compatible and ant then depend on JLine 3.5.2?

**Comment by maggu2810** [comment]
You could also a patched JLine 3.5.1.sp1 version yourself that contains that method.

**Comment by kaikreuzer** [comment]
Created a fixed sp1 branch in a fork https://github.com/openhab/jline3/tree/sp1 and deployed it to https://bintray.com/openhab/mvn/jline/3.5.1-sp1

**Comment by kaikreuzer** [comment]
Trying to create a distro with it now...

**Comment by kaikreuzer** [comment]
Looks promising:
```
Launching the openHAB runtime...

                          __  _____    ____      
  ____  ____  ___  ____  / / / /   |  / __ )     
 / __ \/ __ \/ _ \/ __ \/ /_/ / /| | / __  | 
/ /_/ / /_/ /  __/ / / / __  / ___ |/ /_/ /      
\____/ .___/\___/_/ /_/_/ /_/_/  |_/_____/     
    /_/                        2.2.0-SNAPSHOT
                               - local build -   

Hit '<tab>' for a list of available commands
and '[cmd] --help' for help on a specific command.
Hit '<ctrl-d>' or type 'system:shutdown' or 'logout' to shutdown openHAB.

openhab> list -s -t 0 | grep jline 
 14 │ Active   │  30 │ 3.5.1.sp1              │ org.jline
```
and
```
kai$ ssh -p 8101 openhab@localhost
Password authentication
Password: 

                          __  _____    ____      
  ____  ____  ___  ____  / / / /   |  / __ )     
 / __ \/ __ \/ _ \/ __ \/ /_/ / /| | / __  | 
/ /_/ / /_/ /  __/ / / / __  / ___ |/ /_/ /      
\____/ .___/\___/_/ /_/_/ /_/_/  |_/_____/     
    /_/                        2.2.0-SNAPSHOT
                               - local build -   

Hit '<tab>' for a list of available commands
and '[cmd] --help' for help on a specific command.
Hit '<ctrl-d>' or type 'system:shutdown' or 'logout' to shutdown openHAB.

openhab>                                                                            
```

**Comment by kaikreuzer** [comment]
Anybody wants to try? -> #585

**Comment by kaikreuzer** [comment]
Hopefully now finally fixed by #585. Let's touch wood!

**Comment by wborn** [comment]
This issue is fixed for me with build #1090 ! :+1: 

It does seem like both versions are packaged:

```
$ find . | grep jline
./runtime/system/org/jline
./runtime/system/org/jline/jline
./runtime/system/org/jline/jline/3.5.0
./runtime/system/org/jline/jline/3.5.0/jline-3.5.0.jar
./runtime/system/org/jline/jline/3.5.1-sp1
./runtime/system/org/jline/jline/3.5.1-sp1/jline-3.5.1-sp1.jar
```

Eventhough 3.5.0 is not listed with:

```
openhab> list -s -t 0 |grep jline                                                                                                                                    
 14 │ Active   │  30 │ 3.5.1.sp1              │ org.jline
```

**Comment by kaikreuzer** [comment]
> It does seem like both versions are packaged

That's intended as removing it [isn't a good idea](https://github.com/openhab/openhab-distro/pull/578#issuecomment-346761207) as we have found out.
